gazebo_ros_control now uses joint_limits_interface

jim-rothrock opened the pull request:
I modified gazebo_ros_control so that joint_limits_interface is now used to enforce the limits of effort-controlled joints. The changes in default_robot_hw_sim.cpp require the most recent code in the hydro-devel branch of joint_limits_interface.

jim-rothrock committed:
Added the urdf_string parameter to parseTransmissionsFromURDF().

jim-rothrock committed:
Added the robot_namespace and urdf_model parameters to initSim().

jim-rothrock committed:
Added "joint_limits_interface" and "urdf" to the build dependency list.

jim-rothrock committed:
Additional parameters are passed to robot_hw_sim->initSim(). These parameters
are used by the joint limits interface.

jim-rothrock committed:
Added "joint_limits_interface" and "urdf" to the component list.

jim-rothrock committed:
joint_limits_interface is now used to enforce limits on effort-controlled
joints.

davetcoleman commented on gazebo_ros_control/src/default_robot_hw_sim.cpp:
```diff
@@ -134,8 +147,11 @@ class DefaultRobotHWSim : public gazebo_ros_control::RobotHWSim
         // Create effort joint interface
-        ej_interface_.registerHandle(hardware_interface::JointHandle(
-            js_interface_.getHandle(joint_names_[j]),&joint_effort_command_[j]));
+        const std::string& joint_name = joint_names_[j];
```
do we really need to make a separate variable for joint name? why not just continue to use joint_names_[j] everywhere?

jim-rothrock replied:
If that is the style you are using, I'll change it.

davetcoleman commented on gazebo_ros_control/src/default_robot_hw_sim.cpp:
```diff
@@ -207,12 +227,56 @@ class DefaultRobotHWSim : public gazebo_ros_control::RobotHWSim
+  // Register the limits of the joint specified by joint_name and joint_handle. The limits are retrieved from
+  // joint_limit_nh. If urdf_model is not NULL, limits are retrieved from it also.
+  void registerEffortJointLimits(const std::string& joint_name, const hardware_interface::JointHandle& joint_handle,
+                                 const ros::NodeHandle& joint_limit_nh, const urdf::Model *const urdf_model)
```
line too long!

jim-rothrock replied:
The ROS C++ Style Guide states that the line length limit is 120 characters. Are you using a different limit?

davetcoleman replied:
Yea, you're right. I just have gotten used to so many people following 80 or 100 but I forgot it was 120. Nevermind.

davetcoleman commented on gazebo_ros_control/src/gazebo_ros_control_plugin.cpp:
```diff
@@ -160,8 +162,10 @@ void GazeboRosControlPlugin::Load(gazebo::physics::ModelPtr parent, sdf::Element
+    const urdf::Model *const urdf_model_ptr = urdf_model.initString(urdf_string) ? &urdf_model : NULL;
 
-    if(!robot_hw_sim_->initSim(model_nh_, parent_model_, transmissions_))
+    if(!robot_hw_sim_->initSim(robot_namespace_, model_nh_, parent_model_, urdf_model_ptr, transmissions_))
```
this is good - it only loads the URDF once! it would also be nice for the controller manager to be passed this urdf model ptr and have it pass it along to all loaded controllers - but that is a whole other story (thoughts @adolfo-rt @wmeeusse ?)

adolfo-rt replied:
Previous discussions stated that the ros_control project made the explicit decision to not impose URDF in its core. However, specific controllers and interfaces are free to use URDF when needed.

I mentioned at some point in the past that loading a URDF could be very slow for complex models. I just revisited my previous statement to make it more exact: Loading a complex URDF to the ROS parameter server can indeed be expensive (seconds), but parsing it from the parameter server to a `urdf::Model` data structure is rather fast (tens of milliseconds).

If there are still good arguments for depending on URDF at the core of ros_control, we can discuss it in a separate issue, and evaluate all pros and cons.

davetcoleman replied:
I don't see it as a pressing issue, its just that all of the controllers I've seen do in fact depend on URDF. I hadn't realized there was a decision not to impose URDF in its core.

davetcoleman commented:
Changing the plugin class signature for robot_hw_sim might break any custom plugins in the wild that are using gazebo_ros_control in Hydro - so perhaps we should document this somewhere (like on the wiki [here](http://gazebosim.org/wiki/Tutorials/1.9/ROS_Control_with_Gazebo)). Technically this is an API change that should wait for Indigo... but I don't know if many people are using this functionality yet.

jim-rothrock committed:
Merge remote-tracking branch 'upstream/hydro-devel' into hydro-devel

jim-rothrock committed:
Eliminated a joint_name variable and replaced it with joint_names_[j].
Modified some lines so that they fit in 100 columns. These changes were made
in order to be consistent with the rest of the file.

jim-rothrock commented:
I committed a new version of default_robot_hw_sim.cpp that address concerns made above. The code should be ready to be merged.
